spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb

```diff
@@ -70,5 +75,70 @@ def find_notify(catalog)
         Puppet::Server::ASTCompiler.compile(bolt_request, boltlib_path)
       }.to raise_error(Puppet::Error, /the path to boltlib modules must be provided/)
     end
+
+    context 'when compiling for a plan with fact and plan/target variable collisions' do
+      let(:boltlib_path) { [] }
+
+      it 'properly shadows the relevant variables' do
+        # TODO: This is a temporary stub of the Bolt requires to get this
+        # test to pass. We should figure out how to properly load Bolt code
+        # later for the tests.
+        allow(Puppet::Server::ASTCompiler).to receive(:require).and_call_original
+        allow(Puppet::Server::ASTCompiler).to receive(:require).with(%r{\Abolt/})
+        mock_apply_inventory_class = double('Bolt::ApplyInventory')
+        allow(mock_apply_inventory_class).to receive(:new).with(anything).and_return(double('mock'))
+        stub_const("Bolt::ApplyInventory", mock_apply_inventory_class)
+
+        facts = {
+          "values" => {
+            "plan_var_fact"   => "fact_value",
+            "target_var_fact" => "fact_value",
+          }
+        }
+        variables = {
+          "values" => [
+            # should be shadowed by plan_var_fact fact
+            {"plan_var_fact"  => "plan_value"},
+            {"plan_var_plain" => "plan_value"},
+          ]
+        }
+        target_variables = {
+          "values" => [
+            # should be shadowed by target_var_fact fact
+            {"target_var_fact" => "target_value"},
+            # should be shadowed by plan_var_plain plan variable
+            {"plan_var_plain"  => "target_value"},
+          ]
+        }
+
+        code = <<CODE
```

Comment:
Based on https://github.com/puppetlabs/puppet/blob/main/lib/puppet/pal/pal_impl.rb#L516-L519, I think these cases should test everything.

Reply:
Would there be any reason to cover the case that a var is shadowed more than once (i.e. set in all three of target var/plan var/fact)

Reply:
Would there be any reason to cover the case that a var is shadowed more than once (i.e. set in all three of target var/plan var/fact)

Reply:
No. we have that covered in orch and in bolt.